src/bks_fmt_plug.c

```diff
@@ -0,0 +1,371 @@
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	int index;
+	const int count = *pcount;
+
+	if (any_cracked) {
+		memset(cracked, 0, sizeof(*cracked) * count);
+		any_cracked = 0;
+	}
+
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+#if defined(_OPENMP) || MAX_KEYS_PER_CRYPT > 1
+#endif
+	for (index = 0; index < count; index += MAX_KEYS_PER_CRYPT)
+	{
+		if (cur_salt->format == 0) {
+			unsigned char mackey[20] = { 0 };
+			int mackeylen = cur_salt->hmac_key_size / 8;
+			unsigned char store_hmac_calculated[20];
+
+			pkcs12_pbe_derive_key( 0, cur_salt->iteration_count,
+					MBEDTLS_PKCS12_DERIVE_MAC_KEY,
+					(unsigned char*)saved_key[index],
+					saved_len[index], cur_salt->salt,
+					cur_salt->saltlen, mackey, mackeylen);
+
+			hmac_sha1(mackey, mackeylen, cur_salt->store_data,
+					cur_salt->store_data_length,
+					store_hmac_calculated, 20);
+
+			if (!memcmp(store_hmac_calculated, cur_salt->store_hmac, 20))
+			{
+				cracked[index] = 1;
+#ifdef _OPENMP
+#pragma omp atomic
+#endif
+				any_cracked |= 1;
+			}
+		} else if (cur_salt->format == 1) {
+			unsigned char store_checkum[20];
+			unsigned char iv[16];
+			unsigned char key[32];
+			Twofish_key tkey;
+			int datalen = 0;
+			unsigned char store_data_decrypted[MAX_STORE_DATA_LENGTH];
+			SHA_CTX ctx;
+
+			pkcs12_pbe_derive_key( 0, cur_salt->iteration_count,
+					MBEDTLS_PKCS12_DERIVE_IV,
+					(unsigned char*)saved_key[index],
+					saved_len[index], cur_salt->salt,
+					cur_salt->saltlen, iv, 16);
+			pkcs12_pbe_derive_key( 0, cur_salt->iteration_count,
+					MBEDTLS_PKCS12_DERIVE_KEY,
+					(unsigned char*)saved_key[index],
+					saved_len[index], cur_salt->salt,
+					cur_salt->saltlen, key, 32);
+			Twofish_prepare_key(key, 32, &tkey);
+			datalen = Twofish_Decrypt(&tkey, cur_salt->store_data, store_data_decrypted, cur_salt->store_data_length, iv);
+			if (datalen < 0)
+				continue;
+			SHA1_Init(&ctx);
+			SHA1_Update(&ctx, store_data_decrypted, datalen - 20);
+			SHA1_Final(store_checkum, &ctx);
+
+			if (!memcmp(store_checkum, store_data_decrypted + datalen - 20, 20))
+			{
+				cracked[index] = 1;
+#ifdef _OPENMP
+#pragma omp atomic
+#endif
+				any_cracked |= 1;
+			}
+		}
+	}
+	return count;
+}
+
+static int cmp_all(void *binary, int count)
+{
+	return any_cracked;
+}
+
+static int cmp_one(void *binary, int index)
+{
+	return cracked[index];
+}
+
+static int cmp_exact(char *source, int index)
+{
+	return 1;
+}
+
+static void set_key(char *key, int index)
+{
+	saved_len[index] =
+		strnzcpyn(saved_key[index], key, sizeof(saved_key[index]));
+}
+
+static char *get_key(int index)
+{
+	return saved_key[index];
+}
+
+struct fmt_main fmt_bks = {
+	{
+		FORMAT_LABEL,
+		FORMAT_NAME,
+		ALGORITHM_NAME,
+		BENCHMARK_COMMENT,
+		BENCHMARK_LENGTH,
+		0,
+		PLAINTEXT_LENGTH,
+		BINARY_SIZE,
+		BINARY_ALIGN,
+		SALT_SIZE,
+		SALT_ALIGN,
+		MIN_KEYS_PER_CRYPT,
+		MAX_KEYS_PER_CRYPT,
+		// FMT_NOT_EXACT is required only for BKS keystores
```

Comment:
Does this format support different types and just one of them is FMT_NOT_EXACT? In that case you could opt to clear this flag and then set it on a per-salt basis instead.